prune/prune.go

```diff
@@ -16,184 +16,204 @@ package prune
-type Strategy string
-
-const (
-	// CustomStrategy maximum age of a resource that is desired, Duration
-	CustomStrategy Strategy = "Custom"
-	// MaxAgeStrategy maximum age of a resource that is desired, Duration
-	MaxAgeStrategy Strategy = "MaxAge"
-	// MaxCountStrategy maximum number of a resource that is desired, int
-	MaxCountStrategy Strategy = "MaxCount"
-	// JobKind equates to a Kube Job resource kind
-	JobKind string = "Job"
-	// PodKind equates to a Kube Pod resource kind
-	PodKind string = "Pod"
-)
+/*
+	-------------------------------------------
+	New Auto Pruning API Implementation
+	-------------------------------------------
+*/
+
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	// Registry is the IsPrunableFunc registry
+	Registry *Registry
+
+	// Client is the k8s client that will be used
+	Client client.Client
+
+	// DryRun indicates whether or not we should actually perform pruning or just return the list of pruneable objects
+	// true = Just check, don't prune
+	// false (default) = Prune
+	DryRun bool
+
+	// Strategy is the function used to determine a list of resources that are pruneable
+	Strategy StrategyFunc
 
-// StrategyConfig holds settings unique to each pruning mode
-type StrategyConfig struct {
-	Mode            Strategy
-	MaxAgeSetting   string
-	MaxCountSetting int
-	CustomSettings  map[string]interface{}
+	// Labels is a map of the labels to use for label matching when looking for resources
+	Labels map[string]string
+
+	// Namespace is the namespace to use when looking for resources
+	Namespace string
+
+	// Logger is the logger to use when running pruning functionality
+	Logger logr.Logger
 }
 
-// StrategyFunc function allows a means to specify
-// custom prune strategies
-type StrategyFunc func(ctx context.Context, cfg Config, resources []ResourceInfo) ([]ResourceInfo, error)
-
-// PreDelete function is called before a resource is pruned
-type PreDelete func(ctx context.Context, cfg Config, something ResourceInfo) error
-
-// Config defines a pruning configuration and ultimately
-// determines what will get pruned
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// ErrUnprunable indicates that it is not allowed to prune a specific object.
+type ErrUnprunable struct {
```

Comment:
I wonder if we should drop the `Err` prefix here and just do `Unrpuneable`. Go uses this pattern internally with things like `io.EOF` that aren't necessarily errors, but indicates that an operation cannot continue.

Reply:
I like this idea, especially because of your previous note that it isn't meant to wrap actual errors. I think removing the `Err` prefix would prevent any confusion regarding whether or not this should be treated as an error. Will work on implementing this.